include/eosio/vm/allocator.hpp

```diff
@@ -379,8 +379,8 @@ namespace eosio { namespace vm {
       template <typename T>
       void alloc(size_t size = 1 /*in pages*/) {
          if (size == 0) return;
-         EOS_VM_ASSERT(page != -1, wasm_bad_alloc, "require memory to allocate");
-         EOS_VM_ASSERT(size <= max_pages - page, wasm_bad_alloc, "exceeded max number of pages");
+         EOS_VM_ASSERT(page >= 0, wasm_bad_alloc, "require memory to allocate");
+         EOS_VM_ASSERT(size + page <= max_pages, wasm_bad_alloc, "exceeded max number of pages");
```

Comment:
This change is wrong.  `max_pages - page` is guaranteed to be non-negative, but `size + page` could, in theory, overflow.

Reply:
In theory yes, but it all practical situations this can't.  It would mean that we have allowed so many pages to be allocated that we would have already ran out of memory well before we reached this point.

Reply:
Overflow is potentially caused by a large value of `size` rather than a large value of `page`.  These checks are already redundant with the checks in execution_context, so it doesn't actually matter right now.  If `alloc` were to be called from some future context where they do matter, we shouldn't make assumptions about the value of `size`.

Reply:
Sure, but I viewed that as an even less realistic concern.  That would mean that the caller of alloc directly would be trying to allocate an insane amount of pages.  None of the allocators should be used outside of eos-vm directly, they are not really general purpose designed systems.

The issue is that even if we were to do that, then the problem still remains in your use case of some future context that doesn't have strongly WASM limits.  If we don't check `page` to `max_pages` externally, then we need to check here or else you can underflow and wind up with a bad situation just as easily.  Those particular systems should be more of a design by contract type system to remove as many checks at the base layers as possible.  Higher clients should be able to work within this and make proper inputs.  We don't document anything, and that is a problem so I will concede that.

Reply:
`page <= max_pages` shouldn't need to be checked because it is a class invariant.